sso-proxy: sign upstream requests with a verifiable digital signature

katzdm opened the pull request:
Enable the signing of upstream requests using a digital signature, instead of requiring a shared secret for each upstream service. Work in progress.

jphines commented:
Resolves https://github.com/buzzfeed/sso/issues/16 and potentially resolves https://github.com/buzzfeed/sso/issues/45 dependent on the specific implementation

katzdm commented:
> Resolves #16 and potentially resolves #45 dependent on the specific implementation

I'm looking at addressing these issues separately. This PR will (soon) address #16; issue #45 is forthcoming.

jphines commented on internal/proxy/request_signer.go:
```diff
@@ -0,0 +1,176 @@
+//      other headers in the request are ignored,
+//    <URL> is the string "<PATH>(?<QUERY>)(#FRAGMENT)", where "?<QUERY>" and "#<FRAGMENT>" are
+//      ommitted if the associated components are absent from the request URL,
+//    <BODY> is the body of the Request (may be `nil`; e.g. for GET requests).
```
I have some concerns about signing the body -- particularly if individuals are trying to upload large content files. As it exists today, I believe sso proxy will gracefully handle these cases by streaming the body and not loading them entirely into memory prior to proxying them. 

I think we have to break that behavior here in to support the signing of the body, which would entail us loading the body into memory before proxying the request. 

How important is it to sign the body? Are there other mechanisms we can use to compensate for these risks?

shrayolacrayon replied:
Can we configure signing requests on a per-upstream basis and have it on by default, but for services that may have request bodies that are too large, turn off request signing?

katzdm replied:
@jphines - Currently, the HMAC signing library that we use reads the whole request body into memory, and factors it into the hash. Having said that, I think that the risk you identify is legitimate.

@shrayolacrayon - I think that's a smart idea.

jphines commented on internal/proxy/oauthproxy.go:
```diff
@@ -533,6 +566,13 @@ func (p *OAuthProxy) RobotsTxt(rw http.ResponseWriter, _ *http.Request) {
+// used to sign each request.
+func (p *OAuthProxy) Certs(rw http.ResponseWriter, _ *http.Request) {
+	rw.WriteHeader(http.StatusOK)
+	fmt.Fprint(rw, p.publicCertsJSON)
```
We don't need to convert this to a string to use it here, nor do we _have_ to call WriteHeader, we could just:


```golang
rw.Write(p.RawPublicCerts)
```

Which will default to `200` and write our string as bytes.

katzdm replied:
Nice, good suggestion. Thanks!

jphines commented on internal/proxy/oauthproxy.go:
```diff
@@ -283,6 +291,25 @@ func NewOAuthProxy(opts *Options, optFuncs ...func(*OAuthProxy) error) (*OAuthPr
+	// Also build the `certs` static JSON-string which will be served from a public endpoint.
+	// The key published at this endpoint allows upstreams to decrypt the `Octoboi-Signature`
+	// header, and validate the integrity and authenticity of a request.
+	var certs = map[string]string{}
```
Let's use a struct here instead of string map. We can use struct tags to make the json marshalling easy.

```golang
type Certificates struct {
    PublicKey string `json:signing_key`
}
. . .
certs := &Certificates{PublicKey:opts.RequestSigningKey}
. . .

rawBytes, err := json.MarshalIndent(*certs, "", "  ")
```

katzdm replied:
Good idea.

jphines commented on internal/proxy/oauthproxy.go:
```diff
@@ -533,6 +566,13 @@ func (p *OAuthProxy) RobotsTxt(rw http.ResponseWriter, _ *http.Request) {
+// used to sign each request.
+func (p *OAuthProxy) Certs(rw http.ResponseWriter, _ *http.Request) {
+	rw.WriteHeader(http.StatusOK)
+	fmt.Fprint(rw, p.publicCertsJSON)
```
Alternatively, there are a couple of different styles to these endpoints. I like to crib from Google's oauth2 endpoints so I'd suggest copying either

https://www.googleapis.com/oauth2/v3/certs
```
{
  "keys": [
    {
      "use": "sig",
      "kid": "aa436c3f63b281ce0d976da0b51a34860ff960eb",
      "e": "AQAB",
      "kty": "RSA",
      "alg": "RS256",
      "n": "oGi1WQRrqaSWAC4Ad_urylTi4MDVkxKQBkGzdgWjZhbmOSCkoDXFnMQITBhEneKSNBaV6kVEugWcnqPyeOA3ahD4Aeho7Z3mEaewrDS9f6JyU3CkkhQTgSn7dPqB-ba5dF0mt92ujlV5vo3tEwHrBQLPywBFqb1qv3GDhS2H0sn22NSvlWv6oV9J24CeC2n9gh5NRpHm8leQLgCrfU30uIghSUCDM3TygjMJVxzFrsVSOdDkwyqUEdj4ggG-uSWPXUTT0nGNmh12aPuiBTM_Pj-n8AE3--pc6-7w_aKIx_oLqeYJ8i0SYLKfecTDGJPYrXNXUvYFe7BDPyyYR48lxw"
    },
    {
      "kid": "d1e869e7bf40ddc3d3de08042598ba8350970f0a",
      "e": "AQAB",
      "kty": "RSA",
      "alg": "RS256",
      "n": "nRr1RREJOBHXHZo2jNpJkNXE1faGH1N2T_1Ob1uF84k9prNhdV52e-ul5pzaYTaWQNQFNHqfDa-cBEEh1ytwuU5UjpHG6DTCLbWdz5hu0YjjQ-LJke_8OTxFPDhiTkQ67ZHd_1QGwcXxjvsU_g5m8843Vr0cszErRdY1YclTUMq3HXPDNInMvsl9AJP4w8RGrUsZP0TJk4FLM6quO6igi-Cl7CyHsZ87Y3wC5pQPAKl6Jmo47RDYN-UjRBPeuOZikfRaBqjemMbRt80mrOS0liIk2VKkWyJEkVB07RnsVDptQlZc_neHRDKUdVUxv02ha6h-qHy8YSyQHq10NAeGmw",
      "use": "sig"
    }
  ]
}
```

or the simpler https://www.googleapis.com/oauth2/v1/certs:
```
{
  "d1e869e7bf40ddc3d3de08042598ba8350970f0a": "-----BEGIN CERTIFICATE-----\nMIIDJjCCAg6gAwIBAgIIWnmqQk9sgXYwDQYJKoZIhvcNAQEFBQAwNjE0MDIGA1UE\nAxMrZmVkZXJhdGVkLXNpZ25vbi5zeXN0ZW0uZ3NlcnZpY2VhY2NvdW50LmNvbTAe\nFw0xODExMDkxNDQ5MTJaFw0xODExMjYwMzA0MTJaMDYxNDAyBgNVBAMTK2ZlZGVy\nYXRlZC1zaWdub24uc3lzdGVtLmdzZXJ2aWNlYWNjb3VudC5jb20wggEiMA0GCSqG\nSIb3DQEBAQUAA4IBDwAwggEKAoIBAQCdGvVFEQk4EdcdmjaM2kmQ1cTV9oYfU3ZP\n/U5vW4XziT2ms2F1XnZ766XmnNphNpZA1AU0ep8Nr5wEQSHXK3C5TlSOkcboNMIt\ntZ3PmG7RiOND4smR7/w5PEU8OGJORDrtkd3/VAbBxfGO+xT+DmbzzjdWvRyzMStF\n1jVhyVNQyrcdc8M0icy+yX0Ak/jDxEatSxk/RMmTgUszqq47qKCL4KXsLIexnztj\nfALmlA8AqXomajjtENg35SNEE9645mKR9FoGqN6YxtG3zSas5LSWIiTZUqRbIkSR\nUHTtGexUOm1CVlz+d4dEMpR1VTG/TaFrqH6ofLxhLJAerXQ0B4abAgMBAAGjODA2\nMAwGA1UdEwEB/wQCMAAwDgYDVR0PAQH/BAQDAgeAMBYGA1UdJQEB/wQMMAoGCCsG\nAQUFBwMCMA0GCSqGSIb3DQEBBQUAA4IBAQCM/qA3dh9rikZC4gHNs34MnjFtgrst\nvP/UsBl6or3pu4ILDC4YrQ2WwZ5ONQoS9tLLGCebOXfAhYtkdImivdYjM0ntEfer\nCIynd/kXVTJMrSFPBk0ybu76ZflYZjLZq6HN2Y2f2y1meNQmbIM+Ohn5D6hdOgzp\nO+ukTX5hVA8ADGFaHULfK1xvGl+zIi93jYySO/g3ktUU85R/LTHD3vImiQVOkaIO\n9QoqLa5QG0bBfcspZm8Fqq0NXyR2ZE1iztNHiElfWnxGIUiDdKMZpFwPOaRR3IWn\nEUTC5n7n+Qeyo3rL3iLhC/jn3rouX1FA5J7baL17KzDSiF5eQVlLOIfy\n-----END CERTIFICATE-----\n",
  "aa436c3f63b281ce0d976da0b51a34860ff960eb": "-----BEGIN CERTIFICATE-----\nMIIDJjCCAg6gAwIBAgIILKxOA47kkD0wDQYJKoZIhvcNAQEFBQAwNjE0MDIGA1UE\nAxMrZmVkZXJhdGVkLXNpZ25vbi5zeXN0ZW0uZ3NlcnZpY2VhY2NvdW50LmNvbTAe\nFw0xODExMDExNDQ5MTFaFw0xODExMTgwMzA0MTFaMDYxNDAyBgNVBAMTK2ZlZGVy\nYXRlZC1zaWdub24uc3lzdGVtLmdzZXJ2aWNlYWNjb3VudC5jb20wggEiMA0GCSqG\nSIb3DQEBAQUAA4IBDwAwggEKAoIBAQCgaLVZBGuppJYALgB3+6vKVOLgwNWTEpAG\nQbN2BaNmFuY5IKSgNcWcxAhMGESd4pI0FpXqRUS6BZyeo/J44DdqEPgB6GjtneYR\np7CsNL1/onJTcKSSFBOBKft0+oH5trl0XSa33a6OVXm+je0TAesFAs/LAEWpvWq/\ncYOFLYfSyfbY1K+Va/qhX0nbgJ4Laf2CHk1GkebyV5AuAKt9TfS4iCFJQIMzdPKC\nMwlXHMWuxVI50OTDKpQR2PiCAb65JY9dRNPScY2aHXZo+6IFMz8+P6fwATf76lzr\n7vD9oojH+gup5gnyLRJgsp95xMMYk9itc1dS9gV7sEM/LJhHjyXHAgMBAAGjODA2\nMAwGA1UdEwEB/wQCMAAwDgYDVR0PAQH/BAQDAgeAMBYGA1UdJQEB/wQMMAoGCCsG\nAQUFBwMCMA0GCSqGSIb3DQEBBQUAA4IBAQBtbU+qsKYo0lye/PTwzJstlelbMXFj\nm7TiKfHcVme87nd/EznxVUeWViz7aD7PUl3N+vf69Q+uMln9S+GxGgoYtAUfugOi\n8q28UuDulUHZElXSYB4Ny1dZrP9WMU9eOQaKP7NiZXFCd1jMTJK8WImUVdbujBIP\n1eQNxaoTbyKfLAMDUiVIG6YFEcCt+UeAN0Xc5qHb0X71TPK38zZgCSRxgGtAalso\nQni2r3+HtNjXyqYWXHj6e70invP1vHcuygSAH1sfyUNu8eqmktzgSsndru8eBSmR\n2Yhzoig2Pt18lA+Jh/6WW90o6/VGho8PZfgVHZs7m0uD5VlWCEuxC5/x\n-----END CERTIFICATE-----\n"
}
```

The latter being what have picked for some other internal projects at BuzzFeed but either I think are OK.

jphines commented on internal/proxy/request_signer.go:
```diff
@@ -0,0 +1,176 @@
+}
+
+// Name of the header used to transmit the signature computed for the request.
+var signatureHeader = "Octoboi-Signature"
```
I hate to be a buzzkill, but let's do something a little more professional here. Maybe `SSO-Signature` or even just `Signature`

jphines commented on internal/proxy/oauthproxy.go:
```diff
@@ -211,13 +218,14 @@ func NewRewriteReverseProxy(route *RewriteRoute, config *UpstreamConfig) *httput
+		auth:          config.HMACAuth,
+		cookieName:    opts.CookieName,
+		statsdClient:  opts.StatsdClient,
+		requestSigner: signer,
```
Can we make this optional someone per upstream? I'm still sort of worried about the request body case and want to make it possible to _opt out_ if necessary.

jphines commented on internal/proxy/request_signer.go:
```diff
@@ -0,0 +1,176 @@
+//      hash value received through the `Octoboi-Signature` of the request.
+//
+//  Any requests failing this check should be considered tampered with, and rejected.
+func (signer RequestSigner) Sign(req *http.Request) error {
```
Or should this be a pointer handler and we do have worry about concurrent and racey writes to the hasher?

katzdm replied:
Right, this creates a new object, which I believe ought to avoid racey writes to the `hasher` and `signingKey` fields.

jphines commented:
LGTM

shrayolacrayon commented:
one nit but otherwise this looks great!

katzdm commented on internal/proxy/request_signer.go:
```diff
@@ -0,0 +1,187 @@
+	// Build private key.
+	if block, _ := pem.Decode([]byte(signingKeyPemStr)); block == nil {
+		return nil, fmt.Errorf("could not read PEM block from signing key")
+	} else if key, err := x509.ParsePKCS8PrivateKey(block.Bytes); err != nil {
```
Fair point, done!
